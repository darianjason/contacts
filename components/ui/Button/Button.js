import React from 'react';
import Icon from 'react-native-vector-icons/FontAwesome5';
import {Pressable} from 'react-native';

import {Colors} from '../../../constants';
import styles from './Button.styles';
import DefaultText from '../DefaultText/DefaultText';

const renderIcon = (name, color, size) => (
  <Icon
    name={name}
    color={color ? color : Colors.accent}
    size={size ? size : 18}
  />
);

const renderText = (children, color, name) => (
  <DefaultText
    style={{
      ...styles.text,
      color: color ? color : Colors.accent,
      marginLeft: name && 8,
    }}>
    {children}
  </DefaultText>
);

const Button = ({name, size, color, backgroundColor, onPress, children}) => (
  <Pressable
    onPress={onPress}
    style={({pressed}) => [
      styles.button,
      pressed && styles.pressed,
      children && styles.hasText,
      backgroundColor && {backgroundColor: backgroundColor},
    ]}>
    {name && renderIcon(name, color, size)}
    {children && renderText(children, color, name)}
  </Pressable>
);

export default Button;
