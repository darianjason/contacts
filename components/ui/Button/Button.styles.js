// @flow

import {StyleSheet} from 'react-native';
import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

type Styles = {
  button: ViewStyleProp,
  hasText: ViewStyleProp,
  pressed: ViewStyleProp,
};

const styles: Styles = StyleSheet.create({
  button: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  hasText: {
    flexDirection: 'row',
    paddingVertical: 8,
    paddingHorizontal: 16,
    borderRadius: 32,
  },
  pressed: {
    opacity: 0.8,
  },
});

export default styles;
