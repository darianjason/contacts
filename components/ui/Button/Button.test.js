import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import Button from './Button';

describe('Button component', () => {
  test('can be pressed if not disabled', () => {
    const onPress = jest.fn();

    const {getByText} = render(<Button onPress={onPress}>Test</Button>);

    fireEvent.press(getByText('Test'));
    expect(onPress).toHaveReturnedTimes(1);
  });

  test('cannot be pressed if disabled', () => {
    const onPress = jest.fn();

    const {getByText} = render(
      <Button onPress={onPress} disabled={true}>
        Test
      </Button>,
    );

    fireEvent.press(getByText('Test'));
    expect(onPress).not.toHaveReturned();
  });

  test('renders icon if name is passed', () => {
    const button = renderer.create(<Button name="react" />);
    const icon = button.toJSON().children[0];

    expect(icon).toHaveProperty(
      'props.style[2].fontFamily',
      'FontAwesome5Brands-Regular',
    );
  });

  test('renders text if children is passed', () => {
    const {getByText} = render(<Button>Test</Button>);

    expect(getByText('Test')).toBeTruthy();
  });

  test('renders correctly', () => {
    const tree = renderer
      .create(<Button name="hand-pointer">Button</Button>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
