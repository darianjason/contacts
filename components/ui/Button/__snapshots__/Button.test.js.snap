// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      false,
      Object {
        "borderRadius": 32,
        "flexDirection": "row",
        "paddingHorizontal": 16,
        "paddingVertical": 8,
      },
      undefined,
      undefined,
    ]
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      Array [
        Object {
          "color": "#E2EEF3",
          "fontSize": 18,
        },
        undefined,
        Object {
          "fontFamily": "FontAwesome5Free-Regular",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {
          "fontWeight": "400",
        },
      ]
    }
  >
    ï‰š
  </Text>
  <Text
    style={
      Object {
        "color": "#E2EEF3",
        "fontFamily": "Manrope-Regular",
        "marginLeft": 8,
      }
    }
  >
    Button
  </Text>
</View>
`;
